### PR TITLE
[IMP] pos_self_order: enable kiosk receipt

### PR DESCRIPTION
**Pull request by vlst-odoo**
### [REF] pos*: use /report for qr code generation
pos*: point_of_sale, pos_online_pay

In this pr we remove the custom generation of qr codes and
adapt the codebase to rely on the `/report/barcode` api.

Task: 3550015

### [REF] pos*: simplify receipt printing
pos*: l10n_co_pos, l10n_fr_pos_cert, l10n_gcc_pos,
          l10n_in_pos, l10n_sa_pos, point_of_sale,
          pos_epson_printer, pos_loyalty, pos_mercury,
          pos_online_payment, pos_restaurant, pos_sale

There is a need to print pos receipts in new environments,
such as the kiosk. In this commit we simplify the steps to
generate the receipt data and introduce a portable mechanism
for printing.

Changes:

- removed the `getOrderReceiptEnv` method; this method
  was returning a lot of data that the receipt was not
  actually using; we now simply use the data from
  `Order.export_for_printing`. This means that it will be much
  easier to recreate this data in other environments, such
  as in the kiosk.

- removed the error prone `generate_wrapped_product_name` method.
  It's aim was to split the orderline name, but this task is much
  better accomplished by declarative css.

- removed the `Orderline.export_for_printing` method; instead,
  we simply use the existing `Orderline.getDisplayData`;

- removed the `AbstractReceiptScreen` component and replaced
  it's functionality with the new `printer` service;
    - using this service means that we were free to remove
      the `pos-receipt-print` div from the root of the pos app;

- made the receipt be a standalone component. It thus benefits
  from all the advantages of components, such as prop validation
  and usage of slots;
    - replaced the `OrderLinesReceipt` template with the generic
      `OrderWidget` component'; this greatly simplifies the code;

- created the `renderer` service. It's goal is to do for components
  what `renderToElement` does for templates; we use this service to
  render the receipt component for printing;

- removed jquery from the EpsonPrinter class, such that it can be
  imported into projects that do not rely on jquery;

- created the `ReceiptHeader` component. This unifies the header
  between all the different receipts; ( before, each receipt was
  implementing it's own header )

Task: 3547597

###  [IMP] pos_self_order: enable kiosk receipt
*point_of_sale, pos_self_order, pos_self_order_epson_printer

In this commit we introduce the functionality that allows
the pos_self_order module to generate receipts.

At the moment, this is possible by using Epson Printers via HTTP.
This is done through the use of the newly added bridge module
`pos_self_order_epson_printer`.

In order to generate the data required for the receipt we add more data
on the first load of the kiosk ( such as company information ) and
also in the `_export_for_self_order` method.

Printing is done through the `printer` service from POS.
Task: 3487638
https://github.com/odoo/enterprise/pull/48304

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/point_of_sale/static/src/app/printer/base_printer.js [hunk 1]
```diff
@@ -28,7 +27,8 @@ export class BasePrinter {
         let image, printResult;
         while (this.receiptQueue.length > 0) {
             receipt = this.receiptQueue.shift();
-            image = await this.htmlToImg(receipt);
+            image = this.processCanvas(await toCanvas(receipt, "pos-receipt-print"));
+            receipt.remove();
```
Comment (caburj) [inline addons/point_of_sale/static/src/app/printer/base_printer.js:31]: this is different `receipt` from the original `htmlToImg`. Can't call `remove` method here.

File: addons/point_of_sale/static/src/css/pos_receipts.css [hunk 2]
```diff
@@ -3,7 +3,11 @@
     height: 0;
     overflow: hidden;
     position: absolute;
+    /* Odoo RTL support automatically flip left into right but html2canvas
```
Comment (vlst-odoo) [inline addons/point_of_sale/static/src/css/pos_receipts.css:6]: test this

File: addons/point_of_sale/static/src/app/navbar/sale_details_button/sales_detail_report.xml [hunk 3]
```diff
@@ -3,14 +3,10 @@
 
     <t t-name="point_of_sale.SaleDetailsReport">
         <div class="pos-receipt">
-            <t t-if="pos.company_logo_base64">
-                <img class="pos-receipt-logo" t-att-src="pos.company_logo_base64" alt="Logo"/>
-                <br/>
-            </t>
-            <t t-if="!pos.company_logo_base64" class="pos-receipt-center-align">
+            <object t-attf-data="/web/image?model=res.company&id={{pos.company.id}}&field=logo" type="image">
```
Comment (vlst-odoo) [inline addons/point_of_sale/static/src/app/navbar/sale_details_button/sales_detail_report.xml:6]: test this

File: addons/pos_self_order/static/src/app/models/line.js [hunk 4]
```diff
@@ -43,7 +43,19 @@ export class Line extends Reactive {
...
+        const matches = str.match(regex);
 
+        if (matches && matches.length > 1) {
+            const attributes = matches[matches.length - 1].trim();
+            const productName = str.replace(matches[0], "").trim();
+            return { productName, attributes };
+        }
+
+        return { productName: str, attributes: "" };
+    }
```
Comment (davidmonnom) [inline addons/pos_self_order/static/src/app/models/line.js:58]: Not needed anymore look at `attribute_value_ids` and use attribute formater is needed `attributeFormatter`

File: addons/pos_self_order/static/src/app/models/order.js [hunk 5]
```diff
@@ -34,6 +36,8 @@ export class Order extends Reactive {
         this.lines = order.lines || [];
         this.tracking_number = order.tracking_number || null;
         this.take_away = typeof order.take_away === "boolean" ? order.take_away : null;
+        this.ticketCode = random5Chars(); // 5-digits alphanum code shown on the receipt
```
Comment (davidmonnom) [inline addons/pos_self_order/static/src/app/models/order.js:39]: Field that not in the backend should not be in the models, but if really needed put in under `//data` comment :)

File: addons/l10n_gcc_pos/static/src/overrides/app/screens/receipt_screen/receipt/order_receipt.xml [hunk 6]
```diff
@@ -15,99 +15,58 @@
...
-        <xpath expr="//span[@t-esc='env.utils.formatCurrency(receiptData.receipt.total_rounded)']/.." position="after">
-            <div t-if="receiptData.receipt.is_gcc_country" class="pos-receipt-amount pos-receipt-amount-arabic" t-translation="off">
-                To Pay / للسداد
-                <span t-esc="env.utils.formatCurrency(receiptData.receipt.total_rounded)" class="pos-receipt-right-align"/>
-            </div>
-        </xpath>
-
-        <xpath expr="//span[@t-esc='env.utils.formatCurrency(receiptData.receipt.change)']/.." position="attributes">
-            <attribute name="t-if">!receiptData.receipt.is_gcc_country</attribute>
-        </xpath>
```
Comment (caburj) [inline addons/l10n_gcc_pos/static/src/overrides/app/screens/receipt_screen/receipt/order_receipt.xml:59]: We need to restore this.

File: addons/point_of_sale/static/src/app/navbar/sale_details_button/sales_detail_report.xml (hunk 3 above)
Comment (vlst-odoo) [reply addons/point_of_sale/static/src/app/navbar/sale_details_button/sales_detail_report.xml:6]: the sale report is broken in master so there is no way to test this now

**Comment by vlst-odoo** [comment]
make orderline comp accept other props

File: addons/pos_self_order/models/res_config_settings.py [hunk 7]
```diff
@@ -73,6 +73,12 @@ def custom_link_action(self):
             "views": [[False, "tree"]],
             "domain": [['pos_config_ids', 'in', self.pos_config_id.id]],
         }
+    @api.onchange("pos_self_ordering_service_mode")
+    def _onchange_pos_self_ordering_service_mode(self):
+        return
+        # table_ids = self.pos_floor_ids.table_ids
+        # if self.pos_self_ordering_service_mode == 'table' and self.pos_self_ordering_mode == 'mobile' and not table_ids:
+        #     raise UserError(_("In Self-Order mode, you must have at least one table to use the table service mode"))
```
Comment (caburj) [inline addons/pos_self_order/models/res_config_settings.py:81]: remove this

File: addons/pos_self_order/static/src/app/pages/order_history_page/order_history_page.js [hunk 8]
```diff
@@ -4,16 +4,18 @@ import { Component, useState } from "@odoo/owl";
 import { useSelfOrder } from "@pos_self_order/app/self_order_service";
 import { _t } from "@web/core/l10n/translation";
 import { useService } from "@web/core/utils/hooks";
-
+import { OrderReceipt } from "@point_of_sale/app/screens/receipt_screen/receipt/order_receipt";
 export class OrdersHistoryPage extends Component {
     static template = "pos_self_order.OrdersHistoryPage";
+    static components = { OrderReceipt };
```
Comment (caburj) [inline addons/pos_self_order/static/src/app/pages/order_history_page/order_history_page.js:10]: `OrderReceipt` seems to be unused in this component.

File: addons/point_of_sale/static/src/app/screens/receipt_screen/receipt_screen.js [hunk 9]
```diff
@@ -118,28 +116,42 @@ export class ReceiptScreen extends AbstractReceiptScreen {
...
-        const orderName = order.get_name();
+        const partner = this.currentOrder.get_partner();
         const orderPartner = {
             email: this.orderUiState.inputEmail,
             name: partner ? partner.name : this.orderUiState.inputEmail,
         };
+        await this.sendToCustomer(orderPartner, "action_receipt_to_customer");
+    }
+    async sendToCustomer(orderPartner, methodName) {
+        const ticketImage = await this.renderer.toCanvas(
```
[the remark this answers is not in the document]
Comment (vlst-odoo) [reply addons/point_of_sale/static/src/app/screens/receipt_screen/receipt_screen.js]: fixed :+1:

File: addons/pos_self_order/static/src/app/pages/order_history_page/order_history_page.js (hunk 8 above)
Comment (vlst-odoo) [reply addons/pos_self_order/static/src/app/pages/order_history_page/order_history_page.js:10]: indeed

File: addons/pos_self_order/static/src/app/models/order.js (hunk 5 above)
Comment (vlst-odoo) [reply addons/pos_self_order/static/src/app/models/order.js:39]: yes, i have to save it to the db

**Comment by caburj** [comment]
robodoo rebase-ff

**Pull request closed without merging**